server/app/condition.go

```
@@ -0,0 +1,340 @@
+func (c *Condition) Evaluate(propertyFields []PropertyField, propertyValues []PropertyValue) bool {
+	// fieldID -> PropertyField
+	fieldMap := make(map[string]PropertyField)
+	for _, field := range propertyFields {
+		fieldMap[field.ID] = field
+	}
+
+	// fieldID -> PropertyValue
+	valueMap := make(map[string]PropertyValue)
+	for _, value := range propertyValues {
+		valueMap[value.FieldID] = value
+	}
+
+	return c.evaluate(fieldMap, valueMap)
+}
+
+// Validate ensures the condition is structurally valid and references valid field options
+func (c *Condition) Validate(propertyFields []PropertyField) error {
+	return c.validate(0, propertyFields)
+}
+
+// Sanitize trims whitespace from condition values
+func (c *Condition) Sanitize() {
+	if c.And != nil {
+		for i := range c.And {
+			c.And[i].Sanitize()
+		}
+	}
+
+	if c.Or != nil {
+		for i := range c.Or {
+			c.Or[i].Sanitize()
+		}
+	}
+
+	if c.Is != nil {
+		c.Is.Sanitize()
+	}
+
+	if c.IsNot != nil {
+		c.IsNot.Sanitize()
+	}
+}
+
+func (c *Condition) evaluate(fieldMap map[string]PropertyField, valueMap map[string]PropertyValue) bool {
+	if c.And != nil {
+		for _, condition := range c.And {
+			if !condition.evaluate(fieldMap, valueMap) {
+				return false
+			}
+		}
+		return true
+	}
+
+	if c.Or != nil {
+		for _, condition := range c.Or {
+			if condition.evaluate(fieldMap, valueMap) {
+				return true
+			}
+		}
+		return false
+	}
+
+	if c.Is != nil {
+		field, fieldExists := fieldMap[c.Is.FieldID]
+		if !fieldExists {
+			return false
+		}
+
+		value, valueExists := valueMap[c.Is.FieldID]
+		if !valueExists {
+			return false
+		}
+
+		return is(field, value, c.Is.Value)
+	}
+
+	if c.IsNot != nil {
+		field, fieldExists := fieldMap[c.IsNot.FieldID]
+		if !fieldExists {
+			return true
+		}
+
+		value, valueExists := valueMap[c.IsNot.FieldID]
+		if !valueExists {
+			return true
+		}
+
+		return isNot(field, value, c.IsNot.Value)
+	}
+
+	return true
```

I would have thought that an undefined (presumably future) Condition type would default to false for Evaluate, but I'm don't have a strong opinion - I'd really say this was a program error and should return an error as `validate()` does below.